### PR TITLE
Reduce CPU usage by replacing coll_t by const coll_t&

### PR DESCRIPTION
**Pull request by efirs**
About 9% performance improvement above filestore.

The change is almost trivial replacement of coll_t by const coll_t&, except replacement of call to _kludge_temp_object_collection.

(Recreated PR without move constructor patch, which caused some unittests to fail)

### CODE REVIEW

**Comment by efirs** [comment]
@ldachary looks like this got unnoticed after I recreated PR(I am sorry for that), can you please review one more time?

**Comment by efirs** [comment]
@liewegas Can you review, please?

**Comment by liewegas** [comment]
Actually, I think we should go in a different direction here.  Instead of passing a const coll_t&, we should create an abstract CollectionHandle type, and switch all of the read methods in the ObjectStore interface to take that instead of a coll_t.  For the legacy backends, we can make a handle that just has the coll_t as a single member.  For the new backends, it can have a pointer directly to the implementation's CollectionRef (bluestore, memstore, kstore).

**Comment by efirs** [comment]
@liewegas That's sound like a bigger change, with a different level of risk, I don't feel that familiar with the code yet..

**Comment by tchaikov** [comment]
see #7223.

**Comment by liewegas** [comment]
@efirs can you restrict this patch just to the case in ObjectStore.h's Transaction?  those are unaffected by the collection handle patch

**Comment by efirs** [comment]
@liewegas I can rebase on top of your changes. It'll automatically reject unneeded parts :)

**Comment by liewegas** [comment]
please rebase on #7223

**Comment by liewegas** [comment]
Actually, we should still make the const coll_t switch to the legacy compatibility methods, too.

**Comment by efirs** [comment]
@liewegas Rebased on top of your changes. But I didn't get you last comment, can you rephrase?

**Comment by yuyuyu101** [comment]
LGTM
